### PR TITLE
Daterange test and stricter API use

### PR DESCRIPTION
**Pull request by Avokadoen**
# Changes:
* Parse string values to int before comparing to reduce potential for errors
* Convert date according to pywb [specification](https://pywb.readthedocs.io/en/latest/manual/cdxserver_api.html#from-to) of from and to
* Unit tests for the DateRange struct

### CODE REVIEW

File: pkg/server/warcserver/daterange.go [hunk 1]
```diff
@@ -18,24 +18,130 @@ package warcserver
...
 }
 
-// From pads the timestamp f with 0's on the right until the string is 14 characters in length.
-func From(f string) string {
-	return fmt.Sprintf("%s%0*d", f, 14-len(f), 0)
+// Implemented according to https://pywb.readthedocs.io/en/latest/manual/cdxserver_api.html#from-to
+func From(f string) (int64, error) {
+	fLen := len(f)
+	if fLen%2 != 0 {
+		return 0, fmt.Errorf("'from' string was not an odd number, len: %d", fLen)
```
Comment (maeb) [inline pkg/server/warcserver/daterange.go:65]: Confusing and erroneous error message. If length of input is odd the timestamp cannot be parsed, but the error message claims the string was not odd.

File: pkg/server/warcserver/daterange.go [hunk 2]
```diff
@@ -18,24 +18,130 @@ package warcserver
...
 }
 
-// To pads the timestamp t with 9's on the right until the string is 14 characters in length.
-func To(t string) string {
-	return fmt.Sprintf("%s%.*s", t, 14-len(t), "99999999999999")
+// Implemented according to https://pywb.readthedocs.io/en/latest/manual/cdxserver_api.html#from-to:
+func To(t string) (int64, error) {
+	fLen := len(t)
+	if fLen%2 != 0 {
+		return 0, fmt.Errorf("'to' string was not an odd number, len: %d", fLen)
```
Comment (maeb) [inline pkg/server/warcserver/daterange.go:108]: Error message is wrong.

File: pkg/server/warcserver/daterange.go [hunk 3]
```diff
@@ -18,24 +18,130 @@ package warcserver
...
+	if fLen%2 != 0 {
+		return 0, fmt.Errorf("'from' string was not an odd number, len: %d", fLen)
+	}
+	if fLen > 14 {
+		return 0, fmt.Errorf("expected from string len less than 14, len: %d", fLen)
+	}
+
+	// No specified from date
+	if fLen < 4 {
+		return math.MinInt64, nil
```
Comment (maeb) [inline pkg/server/warcserver/daterange.go:73]:
```suggestion
		return time.Time{}.Unix(), nil
```

File: pkg/server/warcserver/daterange.go [hunk 4]
```diff
@@ -18,24 +18,130 @@ package warcserver
...
+	builder := strings.Builder{}
+	builder.Grow(14)
+	builder.WriteString(t)
+	if builder.Len() <= 4 {
+		builder.WriteString("12")
+	}
+	// Assumption: there is no harm in having a timestamp with a month with less than 31 days
+	// 			   be assigned 31 days
+	if builder.Len() <= 6 {
+		builder.WriteString("31")
```
Comment (maeb) [inline pkg/server/warcserver/daterange.go:128]: The assumption is wrong and will always fail parsing below if month has 28, 29 or 30 days.

Comment (Avokadoen) [reply pkg/server/warcserver/daterange.go]: Made significant changes to parsing. Correctness should be improved and added days should be according to specified month :+1:

**Review by maeb — APPROVED** [review]

**Pull request merged**